saga/src/main/java/com/iluwatar/saga/choreography/Service.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * The MIT License
+ * Copyright © 2014-2019 Ilkka Seppälä
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package com.iluwatar.saga.choreography;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Common abstraction class representing services
+ * implementing a general contract @see {@link Chapter}
+ */
+public abstract class Service implements Chapter {
+    protected static final Logger logger = LoggerFactory.getLogger(Service.class);
+
+    private final ServiceDiscoveryService sd;
+
+    public Service(ServiceDiscoveryService service) {
+        this.sd = service;
+    }
+
+    @Override
+    public Saga execute(Saga saga) {
+        Saga nextSaga = saga;
+        Object nextVal;
+        String chapterName = saga.getCurrent().getName();
+        if (chapterName.equals(getName())) {
+            if (saga.isForward()) {
+                nextSaga = process(saga);
+                nextVal = nextSaga.getCurrentValue();
+                if (nextSaga.isCurrentSuccess()) {
+                    nextSaga.forward();
+                } else {
+                    nextSaga.back();
+                }
+            } else {
+                nextSaga = rollback(saga);
+                nextVal = nextSaga.getCurrentValue();
+                nextSaga.back();
+            }
+
+            if (isSagaFinished(nextSaga)) {
+                return nextSaga;
+            }
+
+            nextSaga.setCurrentValue(nextVal);
+        }
+        Saga finalNextSaga = nextSaga;
+
+        return sd.find(chapterName).map(ch -> ch.execute(finalNextSaga))
+                .orElseThrow(RuntimeException::new);
```

Comment:
Would be nice to have some sort of description in the RuntimeException i.e. "could not execute saga"

Reply:
Agree. done